gapis/gfxapi/vulkan/vulkan_terminator.go

```diff
@@ -0,0 +1,378 @@
+	"github.com/google/gapid/gapis/resolve"
+	"github.com/google/gapid/gapis/service/path"
+)
+
+// VulkanTerminator is very similar to EarlyTerminator.
+// It has 2 additional properties.
+//   1) If a VkQueueSubmit is found, and it contains an event that will be
+//      signaled after the final request, we remove the event from the
+//      command-list, and remove any subsequent events
+//   2) If a request is made to replay until the MIDDLE of a vkQueueSubmit,
+//      then it will patch that command-list to remove any commands after
+//      the command in question.
+//      Furthermore it will continue the replay until that command can be run
+//      i.e. it will make sure to continue to mutate the trace until
+//      all pending events have been successfully completed.
+//      TODO(awoloszyn): Handle #2
+// This takes advantage of the fact that all atoms will be in order.
+type VulkanTerminator struct {
+	lastRequest    atom.ID
+	stopped        bool
+	syncData       *gfxapi.SynchronizationData
+	blockingEvents []VkEvent
+}
+
+func NewVulkanTerminator(ctx context.Context, capture *path.Capture) (*VulkanTerminator, error) {
+	sync, err := database.Build(ctx, &resolve.SynchronizationResolvable{capture})
+	if err != nil {
+		return nil, err
+	}
+	s, ok := sync.(*gfxapi.SynchronizationData)
+	if !ok {
+		return nil, log.Errf(ctx, nil, "Could not get synchronization data")
+	}
+
+	return &VulkanTerminator{atom.ID(0), false, s, []VkEvent(nil)}, nil
+}
+
+// Add adds the atom with identifier id to the set of atoms that must be seen
+// before the VulkanTerminator will consume all atoms (excluding the EOS atom).
+func (t *VulkanTerminator) Add(id atom.ID) {
+	if id > t.lastRequest {
+		t.lastRequest = id
+	}
+}
+
+func walkCommands(s *State,
+	commands CommandBufferCommands,
+	callback func(*CommandBufferCommand)) {
+	for _, c := range commands {
+		callback(&c)
+		if execSub, ok := c.recreateData.(*RecreateCmdExecuteCommandsData); ok {
+			for _, k := range execSub.CommandBuffers.KeysSorted() {
+				cbc := s.CommandBuffers[execSub.CommandBuffers[k]]
+				walkCommands(s, cbc.Commands, callback)
+			}
+		}
+	}
+}
+
+// resolveCurrentRenderPass walks all of the current and pending commands
+// to determine what renderpass we are in after the idx'th subcommand
+func resolveCurrentRenderPass(ctx context.Context, s *gfxapi.State, submit *VkQueueSubmit,
+	idx gfxapi.SubcommandIndex, lrp *RenderPassObject, subpass uint32) (*RenderPassObject, uint32) {
+	if len(idx) == 0 {
+		return lrp, subpass
+	}
+	a := submit
+	c := GetState(s)
+	queue := c.Queues[submit.Queue]
+	l := s.MemoryLayout
+
+	f := func(o *CommandBufferCommand) {
+		switch t := o.recreateData.(type) {
+		case *RecreateCmdBeginRenderPassData:
+			lrp = c.RenderPasses[t.RenderPass]
+			subpass = 0
+		case *RecreateCmdNextSubpassData:
+			subpass += 1
+		case *RecreateCmdEndRenderPassData:
+			lrp = nil
+			subpass = 0
+		}
+	}
+
+	walkCommands(c, queue.PendingCommands, f)
```

Comment:
Why do we walk through the pending commands first, then the not pending ones? It seems that all the results from walking through the pending commands will be overwritten by the following walk through of the not-pending commands.